third_party/move/move-compiler-v2/src/seqs_in_binop_checker.rs

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Pre-requisites: this checker should be run before inlining in its current
+//! incarnation.
+//
+//! This module implements a checker that looks for non-trivial sequences
+//! within binary operations. The v1 compiler's evaluation order semantics
+//! in the presence of sequences within binary operations are not easily
+//! understood or explainable (see examples below).
+//!
+//! Therefore, in compiler v2 (and above), if the language version is less than
+//! 2.0, we will emit an error in such cases. We expect such uses to be rare,
+//! and the user can easily rewrite the code to get explicit evaluation order
+//! that they want.
+//!
+//! In language version 2.0 and above, we will allow sequences within binary
+//! operations, but the evaluation order will be consistently left-to-right,
+//! following the evaluation order semantics used in normal function calls.
+//!
+//! Consider the following examples to see some samples of evaluation order used
+//! by compiler v1 in the presence of sequences within the context of binary
+//! operations. They are meant to showcase how concisely describing the v1 ordering
+//! is hard (as opposed to, a left-to-right evaluation ordering everywhere).
+//!
+//! We number the sub-expressions in their order of their evaluation.
+//! Some (sub-)expressions are left un-numbered if they are irrelevant to the
+//! understanding of the evaluation order.
+//!
+//! case 1: `add` is a user-defined function.
+//! ```move
+//! let x = 1;
+//! add({x = x - 1; x + 8}, {x = x + 3; x - 3}) + {x = x * 2; x * 2}
+//!      ^^^^^^^^^  ^^^^^    ^^^^^^^^^  ^^^^^      ^^^^^^^^^  ^^^^^
```

Comment:
What happens with an "add" function that takes >=3 arguments?  Here the two arguments are being handled differently, but is it the first that is different, or the last, or is behavior more arbitrary?

Reply:
Are you asking here to precisely describe the v1 behavior?

The idea behind adding these examples (which are not meant to be a complete or sufficient) was that the current v1 behavior is hard/unintuitive to describe, so instead of attempting to do so and simulating it, we should instead throw an error.

Reply:
Let me give the answer to your specific question, by providing the evaluation order:
```
let x = 1;
add({x = x - 1; x + 8}, {x = x + 3; x - 3}, {x = x + 100; x / 3}) + {x = x * 2; x * 2}
       1         2           3        4          5          8           6         7
```

Note that things like nesting operators, functions, using short-circuit operations like `&&`, `||`, using control flow redirections like `abort`, `return`, can all affect the evaluation order in v1.